crates/ruff/src/rules/refurb/rules/repeated_append.rs

```diff
@@ -0,0 +1,377 @@
+        format!("Use `{replacement}` instead of repeatedly calling `{name}.append()`")
+    }
+
+    fn autofix_title(&self) -> Option<String> {
+        let RepeatedAppend { replacement, .. } = self;
+        Some(format!("Replace with `{replacement}`"))
+    }
+}
+
+pub(crate) fn convert_consecutive_appends_into_extend(checker: &mut Checker, stmt: &Stmt) {
+    let Some(appends) = match_consecutive_appends(checker, stmt) else {
+        return;
+    };
+
+    // No need to proceed if we have less than 1 append to work with.
+    if appends.len() <= 1 {
+        return;
+    }
+
+    let groups = group_appends(&appends);
+
+    // group borrows from checker, so we can't directly push into checker.diagnostics
+    let mut diagnostics: Vec<Diagnostic> = vec![];
+    diagnostics.reserve_exact(groups.len());
+
+    for group in groups {
+        let replacement = make_suggestion(&group, checker.generator());
+        let start = group.appends.first().unwrap().stmt.start();
+        let end = group.appends.last().unwrap().stmt.end();
+
+        let mut diagnostic = Diagnostic::new(
+            RepeatedAppend {
+                name: group.name().to_string(),
+                replacement: replacement.to_string(),
+            },
+            TextRange::new(start, end),
+        );
+
+        // We only suggest fix when all appends in a group are clumped together.
+        if checker.patch(diagnostic.kind.rule()) && group.is_consecutive {
+            diagnostic.set_fix(Fix::suggested(Edit::replacement(replacement, start, end)));
+        }
+
+        diagnostics.push(diagnostic);
+    }
+    checker.diagnostics.extend(diagnostics);
+}
+
+#[derive(Debug, Clone)]
+struct Append<'a> {
+    receiver: &'a ExprName,
+    binding: &'a Binding<'a>,
+    argument: &'a Expr,
+    stmt: &'a Stmt,
+}
+
+#[derive(Debug)]
+struct AppendGroup<'a> {
+    appends: Vec<Append<'a>>,
+    is_consecutive: bool,
+}
+
+impl<'a> AppendGroup<'a> {
+    fn name(&self) -> &'a str {
+        &self.appends.first().unwrap().receiver.id
+    }
+}
+
+fn match_consecutive_appends<'a>(checker: &'a Checker, stmt: &'a Stmt) -> Option<Vec<Append<'a>>> {
+    // In order to match consecutive statements, we need to go to the tree ancestor of the
+    // given statement, find its position there, and match all 'appends' from there.
+    let siblings: &'a [Stmt] = if checker.semantic().at_top_level() {
+        // If the statement is at the top level, we should go to the parent module.
+        // Module is available in the definitions list.
+        let Definition::Module(ref module_def) =
+            checker.semantic().definitions[DefinitionId::module()]
+        else {
+            return None;
+        };
+        module_def.python_ast
+    } else {
+        // Otherwise, go to the parent, and take its body as a sequence of siblings.
+        let parent = checker.semantic().current_statement_parent()?;
+        traversal::suite(stmt, parent)?
+    };
```

Comment:
It would be nice to find a way to avoid this lookup if the current statement _isn't_ an append, since the iterator at the end will be empty anyway, and I think `traversal::suite` has some cost.

Reply:
I thought about it and didn't figure out an elegant solution to do so. On the other hand, I'm a Rust newbie, so it's very likely that I'm missing it.

Reply:
Hmm... We could perhaps do something like:

```rust
match_append(checker, stmt).map(|append| {
   // Rest of the current function...
   std::iter::once(append).chain(
        siblings
            .iter()
             // Skip the current statement, since we already tested it...
            .skip(stmt_index + 1)
            .map_while(|next| match_append(checker, next))
    ).collect()
}
```

Reply:
That's pretty cool! Thanks